ambry-api/src/main/java/com/github/ambry/account/Dataset.java

```diff
@@ -0,0 +1,176 @@
+   * @param userTags The user defined metadata. Can be null.
+   */
+  Dataset(String accountName, String containerName, String datasetName, VersionSchema versionSchema,
+      Long expirationTimeMs, Integer retentionCount, Map<String, String> userTags) {
+    checkPreconditions(accountName, containerName, datasetName, versionSchema, expirationTimeMs);
+    this.accountName = accountName;
+    this.containerName = containerName;
+    this.datasetName = datasetName;
+    this.versionSchema = versionSchema;
+    this.expirationTimeMs = expirationTimeMs;
+    this.retentionCount = retentionCount;
+    this.userTags = userTags;
+  }
+
+  public enum VersionSchema {
+    TIMESTAMP, MONOTONIC, SEMANTIC
+  }
+
+  /**
+   * @return the account name.
+   */
+  @JsonProperty(ACCOUNT_NAME_KEY)
+  public String getAccountName() {
+    return accountName;
+  }
+
+  /**
+   * @return the container name.
+   */
+  public String getContainerName() {
+    return containerName;
+  }
+
+  /**
+   * @return the dataset name.
+   */
+  public String getDatasetName() {
+    return datasetName;
+  }
+
+  /**
+   * @return the version schema.
+   */
+  public VersionSchema getVersionSchema() {
+    return versionSchema;
+  }
+
+  /**
+   * @return the expiration time in milliseconds since epoch, or -1 if the dataset should be permanent.
+   */
+  public Long getExpirationTimeMs() {
+    return expirationTimeMs;
+  }
+
+  /**
+   * @return the retention of dataset by count. The older versions will be deprecated.
+   */
+  public Integer getRetentionCount() {
+    return retentionCount;
+  }
+
+  /**
+   * @return the user defined metadata.
+   */
+  @JsonProperty(JSON_USER_TAGS_KEY)
+  public Map<String, String> getUserTags() {
+    return userTags;
+  }
+
+  /**
+   * Checks if any required fields is missing for a {@link Dataset} or for any incompatible settings.
+   * @param accountName The name of the account. Cannot be null.
+   * @param containerName The name of the container. Cannot be null.
+   * @param datasetName The name of the dataset. Cannot be null.
+   * @param versionSchema The schema of the version. Cannot be null.
+   * @param expirationTimeMs The expiration time in milliseconds since epoch, or -1 if the dataset should be permanent. Cannot be null.
+   */
+  private void checkPreconditions(String accountName, String containerName, String datasetName,
+      VersionSchema versionSchema, Long expirationTimeMs) {
+    if (accountName == null || containerName == null || datasetName == null || versionSchema == null
+        || expirationTimeMs == null) {
+      throw new IllegalStateException(
+          "At lease one of required fields accountName=" + accountName + " or containerName=" + containerName
+              + " or datasetName=" + datasetName + " or versionSchema=" + versionSchema + " or expirationTimeMs="
+              + expirationTimeMs + " is null");
+    }
+    if (accountName.isEmpty() || containerName.isEmpty() || datasetName.isEmpty()) {
```

Comment:
Shall we check versionSchema.isEmpty since "versionSchema The schema of the version. Cannot be null."?

Reply:
Enums are type safe in java. We can't assign any other value other than specified in enum constants. So I think we don't need to check.